app/controllers/spree/komoju_controller.rb

```diff
@@ -7,18 +7,44 @@ def callback
 
       case params[:type]
       when "payment.captured"
-        order_number = extract_payment_number(params[:data][:external_order_num])
-        payment = Spree::Payment.find_by_number!(order_number)
-        payment.complete! unless payment.completed?
+        payment_captured!
+      when "payment.refunded"
+        payment_refunded!
       end
 
       head 200
     end
 
     private
 
-    def extract_payment_number(external_order_num)
-      external_order_num.split('-').try(:last)
+    def order_number
+      params[:data][:external_order_num].split("-").try(:first)
+    end
+
+    def payment_number
+      params[:data][:external_order_num].split("-").try(:last)
+    end
+
+    def order
+      @order ||= Spree::Order.find_by_number!(order_number)
+    end
+
+    def payment
+      @payment ||= order.payments.find_by_number!(payment_number)
+    end
+
+    def payment_captured!
+      payment.complete! unless payment.completed?
+    end
+
+    def payment_refunded!
+      if payment.completed?
```

Comment:
@rramsden That is a good point.  It wouldn't make an infinite loop as it will still return 200, but the payment stays `complete` even when refunded.

Instead, I improved this logic to check the that refund amount is less than the maximum possible refund value (`credit_allowed`), so that successive calls should not double up refunds.

Do you think I should raise an error if the amount is too large or the currencies disagree?  The `captured` webhook does not raise an error if it is already `complete?`.